File: sdk/src/main/java/com/cxense/cxensesdk/WidgetItemTypeAdapterFactory.java
package com.cxense.cxensesdk;

import com.cxense.cxensesdk.model.WidgetItem;
import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.TypeAdapter;
import com.google.gson.TypeAdapterFactory;
import com.google.gson.reflect.TypeToken;

public class WidgetItemTypeAdapterFactory implements TypeAdapterFactory {
    @Override
    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
        if (!WidgetItem.class.isAssignableFrom(type.getRawType()))
            return null;
        return (TypeAdapter<T>) new WidgetItemTypeAdapter(gson.getAdapter(JsonElement.class));
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/WidgetItemTypeAdapterFactory.kt
package com.cxense.cxensesdk

import com.cxense.cxensesdk.model.WidgetItem
import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.TypeAdapter
import com.google.gson.TypeAdapterFactory
import com.google.gson.reflect.TypeToken

class WidgetItemTypeAdapterFactory : TypeAdapterFactory {
    override fun <T : Any?> create(gson: Gson, type: TypeToken<T>): TypeAdapter<T>? {
        @Suppress("UNCHECKED_CAST")
        return if (WidgetItem::class.java.isAssignableFrom(type.rawType))
            WidgetItemTypeAdapter(
                gson.getAdapter(JsonElement::class.java)
            ) as TypeAdapter<T>?
        else null
    }
}
